feature branch: lint twig files with twigstan

COil opened the pull request:
This branch adds wigstan/twigstan to lint Twig files.


| Q             | A
|---------------| ---
| Branch?       | twigstan
| Cleanup?      | no 
| Bug fix?      | may be
| Fixed tickets | NA
| New feature?  | yes 
| Doc added?    | todo
| Tests pass?   | yes
| Deprecations? | no 
| License       | MIT

COil commented:
18/10/2024 : 17 errors found : 

```
vendor/bin/twigstan                                                                                                                                                   ✔   8.3.11  18:19:58
TwigStan by Ruud Kamphuis and contributors.
Finding dependencies for 8 templates...
Found 0 dependencies...
Compiling 8 templates...
Flattening 8 templates...
Collecting scopes...
 8/8 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

Injecting scope into templates...
Analyzing templates
 8/8 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

Variable 'composer' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_ComposerAction.php:126
🌱 templates/App/Controller/ComposerAction.html.twig:16

Variable 'form' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:148
🌱 templates/App/Controller/FormAction.html.twig:21

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:152
🌱 templates/App/Controller/FormAction.html.twig:28

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:156
🌱 templates/App/Controller/FormAction.html.twig:33

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:160
🌱 templates/App/Controller/FormAction.html.twig:38

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:164
🌱 templates/App/Controller/FormAction.html.twig:43

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:168
🌱 templates/App/Controller/FormAction.html.twig:48

Variable 'dto' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:172
🌱 templates/App/Controller/FormAction.html.twig:53

Variable 'form' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_FormAction.php:183
🌱 templates/App/Controller/FormAction.html.twig:67

Variable 'readme' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_HomeAction.php:126
🌱 templates/App/Controller/HomeAction.html.twig:9

Variable 'form' does not exist.
🔖 offsetAccess.notFound
🐘 compiled__form.php:19
🌱 templates/_form.html.twig:2

Variable 'status_code' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:138
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:5

Variable 'status_code' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:152
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:10

Variable 'status_text' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:156
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:12

Variable 'status_code' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:160
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:12

Variable 'status_code' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:164
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:14

Variable 'status_code' does not exist.
🔖 offsetAccess.notFound
🐘 compiled_error.php:168
🌱 templates/bundles/TwigBundle/Exception/error.html.twig:15

Found 17 errors
```

COil commented:
```
TwigStan by Ruud Kamphuis and contributors.
Finding dependencies for 8 templates...
Found 0 dependencies...
Compiling 8 templates...
Flattening 8 templates...
Collecting scopes...
 8/8 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

Injecting scope into templates...
Analyzing templates
 8/8 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

No errors found
```

ruudk commented on castor.php:
```diff
@@ -205,6 +205,14 @@ function lint_twig(): int
     return exit_code('bin/console lint:twig templates/');
 }
 
+#[AsTask(name: 'twig', namespace: 'stan', description: 'Lint Twig files with twigstan', aliases: ['lint-twigstan'])]
```
This is not about linting. It's static analysis, same as PHPStan.

COil replied:
Humm, I left the same “namespace” as the Symfony command `lint:twig`, but I realize this one is not a really a lint because it checks if the templates are valid and can be compiled, it's more like a static check as well.

COil replied:
So indeed, it may be better to introduce another namespace `static` or `analyse`

ruudk replied:
> it checks if the templates are valid and can be compiled, it's more like a static check as well.

It is **exactly** like PHPStan. It compiles the Twig templates to PHP, then optimizes the templates so that they are a bit easier to analyze, and then invokes PHPStan to analyze the code and report back its errors.

COil replied:
The [Symfony definition](https://symfony.com/blog/new-in-symfony-7-2-translations-linter) does not make the difference between both:

> [Linters](https://en.wikipedia.org/wiki/Lint_(software)) are static code analysis tools used to flag errors like syntactic and stylistic issues. Symfony includes many linter commands to check YAML config files (lint:yaml), container services (lint:container), Twig templates (lint:twig) and even the syntax of XLIFF translation files (lint:xliff). 

So may it is better to keep lint for simplicity.

COil commented on .gitignore:
```diff
@@ -30,3 +30,7 @@
+
+###> twigstan/twigstan ###
+.twigstan
+###< twigstan/twigstan ###
```
Should be added to the bundle recipe @ruudk no?

ruudk replied:
I guess, but there is no recipe yet 😅

COil replied:
Ho sorry, I have forgotten, it is not a bundle yet  (and maybe it is better without).

COil commented on twig-loader.php:
```diff
@@ -0,0 +1,15 @@
+
+(new Dotenv())->bootEnv(__DIR__.'/.env');
+
+$kernel = new Kernel($_ENV['APP_ENV'], (bool) $_ENV['APP_DEBUG']);
```
I had to modify the generated code to make it work. /cc @ruudk

COil commented on twigstan.php:
```diff
@@ -0,0 +1,24 @@
+    ->phpstanConfigurationFile(__DIR__.'/phpstan.neon')
+    ->phpstanMemoryLimit(false)
+    ->twigEnvironmentLoader(__DIR__.'/twig-loader.php')
+    ->twigPaths(__DIR__.'/templates')
```
The default value for `twigPaths` was realy weid, where not putting the default Twig main directory tempate here? /cc @ruudk

COil commented on templates/App/Controller/FormAction.html.twig:
```diff
@@ -1,5 +1,10 @@
 {% extends 'base.html.twig' %}
 
+{% types {
+    dto: 'App\\Dto\\RegisterFormDto',
```
Double `\` won't be necessary with Twig4 @ruudk, right?

ruudk replied:
Correct.

COil commented on src/Kernel.php:
```diff
@@ -5,9 +5,21 @@
 namespace App;
 
 use Symfony\Bundle\FrameworkBundle\Kernel\MicroKernelTrait;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
```
Better to use imports in the generated code @ruudk , it makes the code easier to read IMO.

ruudk replied:
You're right, but how to do that?

The code in question is commented out: 
https://github.com/twigstan/twigstan/blob/1c061819afb537b7d832eafbbd5c6c5bc3363687/example/twig-loader-symfony.php#L15-L26

So doing it like this felt easier , so that people can just copy and paste it. 

I could have also added a comment for the import, but that's extra friction. 

Ideally, this step won't be needed when we have a proper recipe for installation. Maybe a Symfony bundle.

COil replied:
Right. Let's keep this as it is then.

ruudk commented:
After looking at this, I spotted a few bugs/problems in TwigStan that I solved.

I created a PR with updates:
- https://github.com/strangebuzz/MicroSymfony/pull/99

COil commented on templates/App/Controller/FormAction.html.twig:
```diff
@@ -45,7 +45,7 @@
                 <tr>
                     <th>Birthay</th>
-                    <td>{{ dto.birthday|date('Y-m-d') }}</td>
+                    <td>{{ dto.birthday ? dto.birthday|date('Y-m-d') : 'NA' }}</td>
```
Bug found ! 🐛

COil commented:
Hello @ruudk, I have rebased the branch and updated TwigStan and I have many new errors, it is expected? 

* https://github.com/strangebuzz/MicroSymfony/actions/runs/11880748477/job/33104272440?pr=95

ruudk commented:
Did you also update twig/twig?

ruudk commented:
And phpstan/phpstan

COil commented:
> And phpstan/phpstan

Yes, it is PHPStan 2 now!

COil commented:
> Did you also update twig/twig?

* Twig v3.15.0

ruudk commented:
I found the problem. Twig 3.15 was released in the weekend without me knowing it.

I made sure this won't happen again by doing:
- https://github.com/twigstan/twigstan/pull/101
- https://github.com/twigstan/twigstan/pull/102

From now on, TwigStan will support specific / fixed versions of Twig. 

Please update `twigstan/twigstan`.

Btw, PHPStan 2 is not yet supported, it will be soon:
- https://github.com/twigstan/twigstan/pull/96

so it would be best if you could use 1.12 for now.
